File: 270_Closest_Binary_Search_Tree_Value.py
_author_ = 'jake'
_project_ = 'leetcode'

# https://leetcode.com/problems/closest-binary-search-tree-value/
# Given a non-empty binary search tree and a target value, find the value in the BST that is closest to the target.
# Given target value is a floating point.
# You are guaranteed to have only one unique value in the BST that is closest to the target.

# If target is node val then return. Update closest if node val is closer. If target is more than root val then
# no value in left subtree can be closer than node.val so move to right subtree. Vica versa for left subtree.
# Time - O(n)
# Space - O(1)

class Solution(object):
    def closestValue(self, root, target):
        """
        :type root: TreeNode
        :type target: float
        :rtype: int
        """
        closest = root.val

        while root:

            if root.val == target:      # early return since cannot improve
                return root.val

            if abs(root.val - target) < abs(closest - target):
                closest = root.val

            if target < root.val:
                root = root.left
            else:
                root = root.right

        return closest

File: java/270_Closest_Binary_Search_Tree_Value.java
/*
https://leetcode.com/problems/closest-binary-search-tree-value/
Given a non-empty binary search tree and a target value, find the value in the BST that is closest to the target.

Update closest if current node is closer to target then previous closest.  Go left or right according to whether
target is above or below node.  Values in other subtree cannot be closes than current node.
Time - O(n), height of tree
Space - O(1)
*/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

public class Solution {
    public int closestValue(TreeNode root, double target) {

        int closest = root.val;     // assumes root is not null

        while (root != null) {

            if (Math.abs(root.val - target) < Math.abs(closest - target))
                closest = root.val;

            root = target > root.val ? root = root.right : root.left;
        }

        return closest;
    }
}
